source/d3d12/draw_call_tracker.cpp

```diff
@@ -0,0 +1,256 @@
+	{
+		return _depthstencil_resources_by_handle[depthstencilView.ptr];
+	}
+
+	void draw_call_tracker::on_draw(com_ptr<ID3D12Resource> current_depthstencil, UINT vertices)
+	{
+		_global_counter.vertices += vertices;
+		_global_counter.drawcalls += 1;
+
+#if RESHADE_DX12_CAPTURE_DEPTH_BUFFERS
+		com_ptr<ID3D12Resource> targets[D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT];
+
+		if (current_depthstencil == nullptr)
+			// This is a draw call with no depth stencil
+			return;
+
+		std::lock_guard lock(_counters_per_used_depthstencil_mutex);
+
+		if (const auto intermediate_snapshot = _counters_per_used_depthstencil.find(current_depthstencil); intermediate_snapshot != _counters_per_used_depthstencil.end())
+		{
+			intermediate_snapshot->second.stats.vertices += vertices;
+			intermediate_snapshot->second.stats.drawcalls += 1;
+
+			// Find the render targets, if they exist, and update their counts
+			/*for (UINT i = 0; i < D3D12_SIMULTANEOUS_RENDER_TARGET_COUNT; i++)
+			{
+				// Ignore empty slots
+				if (targets[i] == nullptr)
+					continue;
+
+				if (const auto it = intermediate_snapshot->second.additional_views.find(targets[i].get()); it != intermediate_snapshot->second.additional_views.end())
+				{
+					it->second.vertices += vertices;
+					it->second.drawcalls += 1;
+				}
+				else
+				{
+					// This shouldn't happen - it means somehow someone has called 'on_draw' with a render target without calling 'track_rendertargets' first
+					assert(false);
+				}
+			}*/
+		}
+#endif
+	}
+
+#if RESHADE_DX12_CAPTURE_DEPTH_BUFFERS
+	bool draw_call_tracker::check_depthstencil(ID3D12Resource *depthstencil) const
+	{
+		std::lock_guard lock(_counters_per_used_depthstencil_mutex);
+		return _counters_per_used_depthstencil.find(depthstencil) != _counters_per_used_depthstencil.end();
+	}
+	bool draw_call_tracker::check_depth_texture_format(int formatIdx, ID3D12Resource *depthstencil)
+	{
+		assert(depthstencil != nullptr);
+
+		// Do not check format if all formats are allowed (index zero is DXGI_FORMAT_UNKNOWN)
+		if (formatIdx == DXGI_FORMAT_UNKNOWN)
+			return true;
+
+		// Retrieve texture associated with depth stencil
+		D3D12_RESOURCE_DESC desc = depthstencil->GetDesc();
+
+		const DXGI_FORMAT depth_texture_formats[] = {
+			DXGI_FORMAT_UNKNOWN,
+			DXGI_FORMAT_R16_TYPELESS,
+			DXGI_FORMAT_R32_TYPELESS,
+			DXGI_FORMAT_R24G8_TYPELESS,
+			DXGI_FORMAT_R32G8X24_TYPELESS
+		};
+
+		assert(formatIdx > DXGI_FORMAT_UNKNOWN && formatIdx < ARRAYSIZE(depth_texture_formats));
+
+		return make_dxgi_format_typeless(desc.Format) == depth_texture_formats[formatIdx];
+	}
+
+	void draw_call_tracker::track_rendertargets(int format_index, ID3D12Resource *depthstencil, D3D12_CPU_DESCRIPTOR_HANDLE pDepthStencilView)
+	{
+		assert(depthstencil != nullptr);
+
+		if (!check_depth_texture_format(format_index, depthstencil))
+			return;
+
+		std::lock_guard lock(_counters_per_used_depthstencil_mutex);
+
+		if (&_counters_per_used_depthstencil[depthstencil] != nullptr)
```

Comment:
`&_counters_per_used_depthstencil[depthstencil]` can never be `nullptr`, since that is a reference to a struct. So this doesn't make sense?

Reply:
OK Reverted

Reply:
@Crisire, can you help me find a way to switch between the depth buffers withou having to reload the effects ?

Reply:
That's a bit tricky ...

~Is suppose the correct way would mean adding support for dynamic descriptor allocation though. So instead of having a static descriptor heap as part of the effect, there would have to be a way to request an allocator for X number of descriptors, then copy the current effect descriptor table into that new table and then update the entry that references the depth buffer texture to the new texture with `CreateShaderResourceView`. After all that is done the current effect descriptor table would be updated to point to the new one.~

Maybe it works to just keep track where in the descriptor table the depth buffer texture is stored in each effect (aka keep track of the `D3D12_CPU_DESCRIPTOR_HANDLE` of it) and then overwrite that entry for all effects with the new texture via `CreateShaderResourceView`. Have to be careful with synchronization though (this shouldn't happen while there are still frames in flight that may access the descriptor table currently).

Reply:
Thks for your answer. What's sound strange to me is that, if i remove the load_effects calls, the switch works correctly when in depth buffer copy mode, but is does not work when in standard detection mode...